File: CyFile-Android/app/src/test/java/at/tugraz/tc/cyfile/note/InMemoryNoteRepositoryUnitTest.java
package at.tugraz.tc.cyfile.note;

import junit.framework.Assert;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import java.util.HashSet;
import java.util.Set;

import at.tugraz.tc.cyfile.domain.Note;

/**
 * Unit tests for {@link InMemoryNoteRepository}
 * Created by cobri on 3/21/2018.
 */

public class InMemoryNoteRepositoryUnitTest {

    private Set<Note> notes = new HashSet<>();


    private InMemoryNoteRepository inMemoryNoteRepository;

    @Before
    public void setup() {
        inMemoryNoteRepository = new InMemoryNoteRepository(notes);
    }

    @After
    public void tearDown() {
        notes.clear();
    }

    @Test
    public void findAllShouldReturnCorrectNumberOfEntities() {
        notes.add(new Note("id", null, null));

        Assert.assertEquals(1, inMemoryNoteRepository.findAll().size());
    }

    @Test
    public void findOneShouldReturnEntityIfExists() {
        Note n = new Note("my-id23", "name", "content");
        Note n1 = new Note("my-id2", "name2", "content2");
        notes.add(n);
        notes.add(n1);

        Note note = inMemoryNoteRepository.findOne("my-id23");

        Assert.assertEquals(n.getId(), note.getId());
        Assert.assertEquals(n.getTitle(), note.getTitle());
        Assert.assertEquals(n.getContent(), note.getContent());
    }

    @Test
    public void findOneShouldReturnNullIfEntityDoesNotExists() {
        Note n = new Note("my-id", "name", "content");
        Note n1 = new Note("my-id2", "name2", "content2");
        notes.add(n);
        notes.add(n1);

        Note note = inMemoryNoteRepository.findOne("non-existing-id");

        Assert.assertEquals(null, note);
    }

    @Test
    public void saveShouldInsertNonExistingEntity() {
        Note n = new Note(null, "name", "content");
        Note n1 = new Note("my-id2", "name2", "content2");
        notes.add(n1);

        Note saved = inMemoryNoteRepository.save(n);

        //since it is new entity the repository should assign the id
        Assert.assertNotNull(n.getId());
        Assert.assertEquals(n.getTitle(), saved.getTitle());
        Assert.assertEquals(n.getContent(), saved.getContent());

    }

    @Test
    public void saveShouldUpdateExistingEntity() {
        Note n = new Note("my-id1", "name", "content");
        Note n1 = new Note("my-id2", "name2", "content2");
        notes.add(n1);
        notes.add(n);

        n.setTitle("Changed title");
        Note saved = inMemoryNoteRepository.save(n);

        Assert.assertEquals(saved, n);
    }

    @Test(expected = IllegalArgumentException.class)
    public void saveShouldThrowAnErrorIfEntityIsNotValid() {
        Note n = new Note("my-id1", "name", "content");
        Note n1 = new Note("my-id2", "name2", "content2");
        notes.add(n1);
        notes.add(n);

        inMemoryNoteRepository.save(null);

    }

    @Test
    public void deleteShouldDeleteEntityIfExists() {
        Note n = new Note("my-id1", "name", "content");
        Note n1 = new Note("my-id2", "name2", "content2");
        notes.add(n1);
        notes.add(n);

        inMemoryNoteRepository.delete("my-id1");

        Assert.assertTrue(!notes.contains(n));
        Assert.assertTrue(notes.size() == 1);
    }

    @Test
    public void deleteShouldNotDeleteEntityIfNotExists() {
        Note n = new Note("my-id1", "name", "content");
        Note n1 = new Note("my-id2", "name2", "content2");
        notes.add(n1);
        notes.add(n);

        inMemoryNoteRepository.delete("non-existing-id");

        Assert.assertTrue(notes.contains(n));
        Assert.assertTrue(notes.contains(n1));
        Assert.assertTrue(notes.size() == 2);
    }


}

File: CyFile-Android/app/src/test/kotlin/at/tugraz/tc/cyfile/note/InMemoryNoteRepositoryUnitTest.kt
package at.tugraz.tc.cyfile.note

import at.tugraz.tc.cyfile.BaseUnitTest
import at.tugraz.tc.cyfile.domain.Note
import at.tugraz.tc.cyfile.note.impl.InMemoryNoteRepository
import org.junit.Assert
import org.junit.After
import org.junit.Before
import org.junit.Test
import java.util.*

/**
 * Unit tests for [InMemoryNoteRepository]
 * Created by cobri on 3/21/2018.
 */

class InMemoryNoteRepositoryUnitTest : BaseUnitTest() {

    private val notes = HashSet<Note>()


    private lateinit var inMemoryNoteRepository: InMemoryNoteRepository

    @Before
    fun setup() {
        inMemoryNoteRepository = InMemoryNoteRepository(notes)
    }

    @After
    fun tearDown() {
        notes.clear()
    }

    @Test
    fun findAllShouldReturnCorrectNumberOfEntities() {
        notes.add(Note("id", "", ""))
        Assert.assertEquals(1, inMemoryNoteRepository.findAll().size)
    }

    @Test
    fun findOneShouldReturnEntityIfExists() {
        val n = Note("my-id23", "name", "content")
        val n1 = Note("my-id2", "name2", "content2")
        notes.add(n)
        notes.add(n1)

        val note = inMemoryNoteRepository.findOne("my-id23")

        Assert.assertEquals(n.id, note!!.id)
        Assert.assertEquals(n.title, note.title)
        Assert.assertEquals(n.content, note.content)
    }

    @Test
    fun findOneShouldReturnNullIfEntityDoesNotExists() {
        val n = Note("my-id", "name", "content")
        val n1 = Note("my-id2", "name2", "content2")
        notes.add(n)
        notes.add(n1)

        val note = inMemoryNoteRepository.findOne("non-existing-id")

        Assert.assertEquals(null, note)
    }

    @Test
    fun saveShouldInsertNonExistingEntity() {
        val n = Note("id1", "name", "content")
        val n1 = Note("my-id2", "name2", "content2")
        notes.add(n1)

        val saved = inMemoryNoteRepository.save(n)

        //since it is new entity the repository should assign the id
        Assert.assertNotNull(n.id)
        Assert.assertEquals(n.title, saved.title)
        Assert.assertEquals(n.content, saved.content)

    }

    @Test
    fun saveShouldUpdateExistingEntity() {
        val n = Note("my-id1", "name", "content")
        val n1 = Note("my-id2", "name2", "content2")
        notes.add(n1)
        notes.add(n)

        n.title = "Changed title"
        val saved = inMemoryNoteRepository.save(n)

        Assert.assertEquals(saved, n)
    }

    @Test
    fun deleteShouldDeleteEntityIfExists() {
        val n = Note("my-id1", "name", "content")
        val n1 = Note("my-id2", "name2", "content2")
        notes.add(n1)
        notes.add(n)

        inMemoryNoteRepository.delete("my-id1")

        Assert.assertTrue(!notes.contains(n))
        Assert.assertTrue(notes.size == 1)
    }

    @Test
    fun deleteShouldNotDeleteEntityIfNotExists() {
        val n = Note("my-id1", "name", "content")
        val n1 = Note("my-id2", "name2", "content2")
        notes.add(n1)
        notes.add(n)

        inMemoryNoteRepository.delete("non-existing-id")

        Assert.assertTrue(notes.contains(n))
        Assert.assertTrue(notes.contains(n1))
        Assert.assertTrue(notes.size == 2)
    }


}
